# Digital Signal Processing
## Convolution and Correlation

### MsCV - ViBOT - MAIA

In [1]:
%matplotlib inline
%pprint off

# To make some nice plot
import matplotlib.pyplot as plt
# To play with arrays
import numpy as np
from skimage import io
from skimage import img_as_float
from matplotlib import cm

Pretty printing has been turned OFF


## Exercise 1- 1D Convolution

**1.1** Defining a signal $x$ and an impulse response $h$ as parameters, write a program `conv_fn` to compute the convolution product $y=h*x$. Do not use the python or matlab function and just consider zero padding to solve the border problem.

**1.2** Test your function, while using the following signal ($x[n] = \{\underline{1},2,3,4\}$) and following filters, ($\delta[k]$, $H[k]$, $e^{k}$, and $h = \{-1, 1\}$), respectively.

**1.3** Consider the previous input signal, write it's extended version while considering `symmetry`, `periodic` and `constant values` padding and $h = H[k-5]$



In [41]:
def conv_fn(x, h):

    end = len(x) + 2*len(h)

    X_pad = np.zeros(end)
    X_pad[0:len(x)] = x[0:len(x)]
    X_pad[len(x):end] = 0
    y = np.zeros((end, ))
    
    print X_pad
    print h

    #x[len(x):len(x)+len(h)] = 0
    #x[-1:-1-len(h)] = 0
    
    for i in range(0, len(X_pad)):
        for j in range (0, len(h)):
           y[i] = y[i] + X_pad[i-j]* h[j]
    print y
    return y

In [51]:
def dirac(n=0, N=20):
    # Check that the parameter are valid
    if n > N - 1:
        raise ValueError('n should be inferior to N-1')
    
    # Create the array
    s = np.zeros((N, ))
    # Place the dirac
    s[n] = 1
    
    return s

def step(n=0, N=20):
    # Check that the parameter are valid
    if n > N - 1:
        raise ValueError('n should be inferior to N-1')
    
    # Create the array
    s = np.zeros((N, ))
    # Create the ramp
    for idx_s in np.arange(n, N):
        s[idx_s] = 1
    
    return s

def exp(n=0, N=20):
    # Check that the parameter are valid
    if n > N - 1:
        raise ValueError('n should be inferior to N-1')
    
    # Create the array
    s = np.zeros((N, ))
    e = 2.71828
    # Create the ramp
    for incr, idx_s in enumerate(np.arange(n, N)):
        s[idx_s] = e ** incr
    
    return s



x = np.array([1, 2, 3, 4])
#h = np.array([-1, 1])
h = exp(0)
y = conv_fn(x, h)
print np.convolve(x, h)

[ 1.  2.  3.  4.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.]
[  1.00000000e+00   2.71828000e+00   7.38904616e+00   2.00854964e+01
   5.45980031e+01   1.48412660e+02   4.03427165e+02   1.09662799e+03
   2.98094195e+03   8.10303487e+03   2.20263176e+04   5.98736987e+04
   1.62753478e+05   4.42409523e+05   1.20259296e+06   3.26898439e+06
   8.88601488e+06   2.41546765e+07   6.56591741e+07   1.78480020e+08]
[  1.00000000e+00   4.71828000e+00   1.58256062e+01   4.70184287e+01
   1.27809254e+02   3.47421340e+02   9.44388480e+02   2.56711232e+03
   6.97813007e+03   1.89685114e+04   5.15617252e+04   1.40159206e+05
   3.80991967e+05   1.03564285e+06   2.81516723e+06   7.65241279e+06
   2.08014006e+07   5.65440313e+07   1.53702509e+08   4.17806457e+08
   6.50556268e+08   7.98076756e+08   7.13920080e+08   0.00000000e+00
   0.00000000e+00   0.00000000e+00   0.00000000e+00   0.0000

## Exercise 2- 2D Convolution

**2.1** Define a 2D convolution function (Only consider zero-padding the image)

**2.2** Using your function, smooth the `lena` image with the following Gaussian kernel:
$$ K = \frac{1.}{256}\begin{bmatrix} 1 & 4 & 6 & 4 & 1 \\ 4 & 16 & 24 & 16 & 4 \\ 6 & 24 & 36 & 24 & 6 \\ 4 & 16 & 24 & 16 & 4 \\ 1 & 4 & 6 & 4 & 1 \end{bmatrix}$$

**2.3** Apply sobel filtering using convolution


## Exercise 3 - Character recognition using cross-correlation

**3.1** load `a.png` and `text.png` images and binarized them using [`threshold_otsu`](http://scikit-image.org/docs/dev/api/skimage.filters.html?highlight=filters#skimage.filters.threshold_otsu)

**3.2** Use [`signal.correlate2d`](https://docs.scipy.org/doc/scipy-0.18.1/reference/generated/scipy.signal.correlate2d.html#scipy.signal.correlate2d) from `scipy` to correlate your binarized images.
**Note, signal correspond to the bright pixels within and image (invert your masks)**

**3.3** Find the first 10 characters recognized as `a`.